zkevm-circuits/src/state_circuit/lexicographic_ordering.rs

```diff
@@ -0,0 +1,298 @@
+                diff_selector * diff_1,
+                meta.query_fixed(u16_range, Rotation::cur()),
+            )]
+        });
+        meta.lookup_any("diff_2 fits into u16", |meta| {
+            let diff_selector = meta.query_advice(diff_selector, Rotation::cur());
+            let diff_2 = meta.query_advice(diff_2, Rotation::cur());
+            vec![(
+                not::expr(diff_selector) * diff_2,
+                meta.query_fixed(u16_range, Rotation::cur()),
+            )]
+        });
+
+        meta.create_gate("diff_selector is boolean", |meta| {
+            let selector = meta.query_fixed(selector, Rotation::cur());
+            let diff_selector = meta.query_advice(diff_selector, Rotation::cur());
+            vec![selector * diff_selector.clone() * not::expr(diff_selector)]
+        });
+
+        meta.create_gate("diff_inverse", |meta| {
+            let selector = meta.query_fixed(selector, Rotation::cur());
+            let diff_selector = meta.query_advice(diff_selector, Rotation::cur());
+            let diff_inverse = meta.query_advice(diff_inverse, Rotation::cur());
+            let diff_1 = meta.query_advice(diff_1, Rotation::cur());
+            let diff_2 = meta.query_advice(diff_2, Rotation::cur());
+            vec![
+                selector
+                    * select::expr(
+                        diff_selector,
+                        diff_inverse.clone() * diff_1 - 1u64.expr(),
+                        diff_inverse * diff_2 - 1u64.expr(),
+                    ),
+            ]
+        });
+
+        config
+    }
+}
+
+struct Queries<F: Field> {
+    tag: Expression<F>,       // 4 bits
+    field_tag: Expression<F>, // 8 bits, so we can pack tag + field_tag into one limb.
+    id_limbs: [Expression<F>; N_LIMBS_ID],
+    address_limbs: [Expression<F>; N_LIMBS_ACCOUNT_ADDRESS],
+    storage_key_bytes: [Expression<F>; N_BYTES_WORD],
+    rw_counter_limbs: [Expression<F>; N_LIMBS_RW_COUNTER],
+}
+
+impl<F: Field> Queries<F> {
+    fn new(meta: &mut VirtualCells<'_, F>, config: Config, rotation: Rotation) -> Self {
+        let mut query_advice = |column| meta.query_advice(column, rotation);
+        Self {
+            // witness: query_advice(config.witness),
+            tag: query_advice(config.tag),
+            field_tag: query_advice(config.field_tag),
+            id_limbs: config.id_limbs.map(&mut query_advice),
+            address_limbs: config.address_limbs.map(&mut query_advice),
+            storage_key_bytes: config.storage_key_bytes.map(&mut query_advice),
+            rw_counter_limbs: config.rw_counter_limbs.map(query_advice),
+        }
+    }
+
+    fn packed_tags(&self) -> Expression<F> {
+        (1u64 << 4).expr() * self.tag.clone() + self.field_tag.clone()
+    }
+
+    fn storage_key_limbs(&self) -> Vec<Expression<F>> {
+        self.storage_key_bytes
+            .iter()
+            .tuples()
+            .map(|(hi, lo)| (1u64 << 16).expr() * hi.clone() + lo.clone())
```

Comment:
1u64 << 8
besides, i think the endianness is wrong.   storage_key_bytes is little endian, so `map(|(lo, hi)| ...`?